### PR TITLE
@okyeron's MI-Ugens pedals

### PR DESCRIPTION
**Pull request by 21echoes**

### CODE REVIEW

File: lib/fx/Clouds.sc [hunk 1]
```diff
@@ -0,0 +1,19 @@
+CloudsPedal : Pedal {
+  *id { ^\clouds; }
+
+  *fxArguments { ^[\pit, \pos, \size, \dens, \tex, \spread, \mode, \fb ]; }
+
+  *fxDef {^{|wet|
+    var pit=0, pos=0.5, size=0.5, dens=0.5, tex=0, drywet=1, in_gain=1, spread=0.5, rvb=0.5, fb=0, freeze=0, mode=0, lofi=0, trig=0;
+	pit = \pit.kr(0.7);
```
Comment (21echoes) [inline lib/fx/Clouds.sc]: some of these defaults are a little confusing to me -- e.g. why does pitch start at `0.7`?

File: lib/ui/pedals/clouds.lua [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    id = id_prefix .. "_pit",
+    name = "Pitch",
+    type = "control",
+    controlspec = ControlSpec.new(-48, 48, "lin", 1, 0, ""),
+  }
+  local pos_control = {
+    id = id_prefix .. "_pos",
+    name = "Position",
+    type = "control",
+    controlspec = ControlSpec.new(0.00, 1.00, "lin", 0.01, 0.5, ""),
```
Comment (21echoes) [inline lib/ui/pedals/clouds.lua]: I probably shouldn't have done this, but seeing as I have.... `Controlspecs.MIX` is the preferred way in this app to have a 0-100 range input that is sent to SC as 0-1 range
```suggestion
    controlspec = Controlspecs.MIX
```

File: lib/ui/pedals/clouds.lua [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+end
+
+function CloudsPedal.params()
+  local id_prefix = CloudsPedal.id
+
+  local pitch_control = {
+    id = id_prefix .. "_pit",
+    name = "Pitch",
+    type = "control",
+    controlspec = ControlSpec.new(-48, 48, "lin", 1, 0, ""),
```
Comment (21echoes) [inline lib/ui/pedals/clouds.lua:39]: unit "st" is standard for semitones throughout the app

File: lib/ui/pedals/clouds.lua [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    name = "Spread",
+    type = "control",
+    controlspec = ControlSpec.new(0.00, 1.00, "lin", 0.01, 0, ""),
+  }
+
+  local mode_control = {
+    id = id_prefix .. "_mode",
+    name = "Mode",
+    type = "control",
+    controlspec = ControlSpec.new(0, 3, "lin", 1, 0, ""),
```
Comment (21echoes) [inline lib/ui/pedals/clouds.lua]: I think a generic "0, 1, 2, 3" will be confusing -- let's convert this to a `options` instead of `control` and name the options. The `Pedal` superclass will handle converting to a 0-indexed number for us :+1:

File: lib/ui/pedals/clouds.lua [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    name = "Density",
+    type = "control",
+    controlspec = ControlSpec.new(0.00, 1.00, "lin", 0.01, 0.5, ""),
+  }
+
+  local tex_control = {
+    id = id_prefix .. "_tex",
+    name = "Texture",
+    type = "control",
+    controlspec = ControlSpec.new(0.00, 1.00, "lin", 0.01, 0, ""),
```
Comment (21echoes) [inline lib/ui/pedals/clouds.lua]: for 0-100 values sent to SC as 0-1 values with non-50/0.5 defaults, there's `Controlspecs.mix(default)`
```suggestion
    controlspec = Controlspecs.mix(0)
```

File: lib/ui/pedals/rings.lua [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    {"Position", "Model/EasterEgg"},
+    i:_default_section(),
+  }
+  i:_complete_initialization()
+
+  return i
+end
+
+function RingsPedal:name(short)
+  return short and "RNGS" or "Rings"
```
Comment (21echoes) [inline lib/ui/pedals/rings.lua:29]: worried about confusion with "Ring Mod" 🤔 i'll think on it for a bit

File: lib/ui/pedals/rings.lua [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+end
+
+function RingsPedal.params()
+  local id_prefix = RingsPedal.id
+
+  local pitch_control = {
+    id = id_prefix .. "_pit",
+    name = "Pitch",
+    type = "control",
+    controlspec = ControlSpec.new(0, 127, "lin", 1, 60, ""),
```
Comment (21echoes) [inline lib/ui/pedals/rings.lua]: in my experience, any control with a range larger than 100 possible values doesn't play nice with norns' encoders (some values are skipped). Elsewhere in the app I've narrowed the range to 24-103 🤷 
```suggestion
    controlspec = ControlSpec.new(24, 103, "lin", 1, 60, ""), -- c3 by default
```

File: lib/ui/pedals/rings.lua [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    name = "Position",
+    type = "control",
+    controlspec = ControlSpec.new(0.00, 1.00, "lin", 0.01, 0.2, ""),
+  }
+
+  local model_control = {
+    id = id_prefix .. "_model",
+    name = "Model",
+    type = "control",
+    controlspec = ControlSpec.new(0, 5, "lin", 1, 0, ""),
```
Comment (21echoes) [inline lib/ui/pedals/rings.lua]: Here we should also use `type = "option"` with named options instead of an anonymous "control"

File: lib/ui/pedals/rings.lua [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  local easteregg_control = {
+    id = id_prefix .. "_easteregg",
+    name = "EasterEgg",
+    type = "control",
+    controlspec = ControlSpec.new(0, 1, "lin", 1, 0, ""),
```
Comment (21echoes) [inline lib/ui/pedals/rings.lua]: this should also be a `type = "option"` with `{"On", "Off"}`

**Review by 21echoes — COMMENTED** [review]
@okyeron i can make most all of these changes (and a few nitpicks more, probably 🤐), so no worries. just wanted to document my thoughts and reasoning! lmk if you have any questions. And sorry about the delay! I've been without my laptop for a little while and just got set up again.

One other thing: I also try to do SC CPU testing (Use `fork { loop { [context.server.peakCPU, context.server.avgCPU].postln; 3.wait } };` in `Engine_Pedalboard.alloc`) for all these and make sure the CPU isn't too obscene. Honestly ideally it'd stay under 20% with just this pedal active (the baseline code clocks in at 8%, so 3 pedals at ~12%, plus the reverb pedal which is like 30%, gets you to 75%, which sadly can drop samples on rare occasions. any higher and it starts really dropping samples)

And finally: let me know what results you find from your experiments in distribution! https://llllllll.co/t/mi-ugens-for-norns/31781/42

tl;dr:
- [x] make some changes to make it fit in with the rest of the existing code
- [x] some CPU testing to make sure these aren't too expensive
- [x] figure out how to make it possible to distribute the compiled UGens (and add them to this PR)

File: lib/fx/Clouds.sc (hunk 1 above)
Comment (okyeron) [reply lib/fx/Clouds.sc]: I probably don't know what I'm doing. Is that a starting value or a default value? (are those different things?)

I was thinking they were just some defaults, so I threw some from some of the helpfile examples.

**Comment by okyeron** [comment]
generally this all sounds good. I'm still new to the controlspec stuff, so "option" type and the MIX thing are not something I'd run into yet.

File: lib/ui/pedals/rings.lua (hunk 6 above)
Comment (okyeron) [reply lib/ui/pedals/rings.lua:29]: miRings?

File: lib/ui/pedals/rings.lua (hunk 7 above)
Comment (okyeron) [reply lib/ui/pedals/rings.lua]: controlspec.MIDINOTE ?

File: lib/ui/pedals/clouds.lua (hunk 3 above)
Comment (okyeron) [reply lib/ui/pedals/clouds.lua:39]: FWIW - i'm not sure the value is actually semitones. Manual does not say and I'm not sure where to dig for that in the source

**Comment by okyeron** [comment]
pushed changes based on the notes - will test shortly

**Comment by 21echoes** [comment]
Closing in favor of https://github.com/21echoes/pedalboard/pull/2
